pyomo/gdp/disjunct.py

```diff
@@ -74,28 +76,43 @@ def process(arg):
 
 class _DisjunctData(_BlockData):
 
+    @property
+    def transformation_block(self):
+        return self._transformation_block
+
     def __init__(self, component):
         _BlockData.__init__(self, component)
         self.indicator_var = Var(within=Binary)
+        # pointer to transformation block if this disjunct has been
+        # transformed. None indicates it hasn't been transformed.
+        self._transformation_block = None
 
     def set_value(self, val):
         _indicator_var = self.indicator_var
-        # Remove everything
-        for k in list(getattr(self, '_decl', {})):
-            self.del_component(k)
-        self._ctypes = {}
-        self._decl = {}
-        self._decl_order = []
-        # Now copy over everything from the other block.  If the other
-        # block has an indicator_var, it should override this block's.
-        # Otherwise restore this block's indicator_var.
+        # if _transformation_block is not None we should yell. else call base
+        # class, check that you have indicator_var and transformation block.
+        _transformation_block = self._transformation_block
+        if not _transformation_block is None:
+            # the disjunct has been transformed, so the truth no longer lives
+            # here and this is a bad idea!
+            raise GDP_Error("Attempting to call set_value on an already-"
+                            "transformed disjunct! Since disjunct %s "
+                            "has been transformed, replacing it here will "
+                            "not affect the model." % self.name)
+
+        # Call the base class set_value. (Note we are changing decl_order by
+        # doing things this way: the indicator var will always be last rather
+        # than first)
+        super(_DisjunctData, self).set_value(val)
+
         if val:
             if 'indicator_var' not in val:
                 self.add_component('indicator_var', _indicator_var)
-            for k in sorted(iterkeys(val)):
-                self.add_component(k,val[k])
+            if 'transformation_block' not in val:
+                self._transformation_block = _transformation_block
         else:
             self.add_component('indicator_var', _indicator_var)
+            self._transformation_block = _transformation_block 
```

Comment:
Also, I am confused: wouldn't the value of `_transformation_block` always be `None` here, given that an error would have been raised above otherwise?